networkx/algorithms/d_separation.py

```diff
@@ -140,3 +140,202 @@ def d_separated(G, x, y, z):
+
+    # find a separating set Z' in moral_G
+    Z_prime = set(G.predecessors(x)).union(set(G.predecessors(y)))
+
+    # perform BFS on the graph from 'x' to mark
+    Z_dprime = _bfs_with_marks(moral_G, x, Z_prime)
+    Z = _bfs_with_marks(moral_G, y, Z_dprime)
+    return Z
+
+
+@not_implemented_for("undirected")
+def is_minimal_d_separator(G, x, y, z):
+    """Determine if a d-separating set is minimal.
+
+    Uses the algorithm 2 presented in [1]_. The complexity of the algorithm
+    is :math:`O(|E_{An}^m|)`, where :math:`E_{An}^m` stands for the
+    number of edges in the moralized graph of the sub-graph consisting
+    of only the ancestors of X and Y. For full details, see [1]_.
+
+    Parameters
+    ----------
+    G : nx.DiGraph
+        The graph.
+    x : node
+        X node.
+    y : node
+        Y node.
+    z : Set
+        The separating set to check is minimal.
+
+    Returns
+    -------
+    bool
+        Whether or not the `z` separating set is minimal.
+
+    Raises
+    ------
+    NetworkXError
+        The *d-separation* test is commonly used with directed
+        graphical models which are acyclic.  Accordingly, the algorithm
+        raises a :exc:`NetworkXError` if the input graph is not a DAG.
+
+    NodeNotFound
+        If any of the input nodes are not found in the graph,
+        a :exc:`NodeNotFound` exception is raised.
+
+    References
+    ----------
+    .. [1] Tian, J., & Paz, A. (1998). Finding Minimal D-separators.
+
+    Notes
+    -----
+    This function only works on verifying a d-separating set is minimal
+    between two nodes. To verify that a d-separating set is minimal between
+    two sets of nodes is not supported.
+    """
+    if not nx.is_directed_acyclic_graph(G):
+        raise nx.NetworkXError("graph should be directed acyclic")
+
+    union_xy = set(x).union(set(y)).union(z)
+
+    if any(n not in G.nodes for n in union_xy):
+        raise nx.NodeNotFound("one or more specified nodes not found in the graph")
+
+    x_anc = nx.ancestors(G, x)
+    y_anc = nx.ancestors(G, y)
+    xy_anc = x_anc.union(y_anc)
+
+    # if Z contains any node which is not in ancestors of X or Y
+    # then it is definitely not minimal
+    if any(node not in xy_anc for node in z):
+        return False
+
+    D_anc_xy = x_anc.union(y_anc)
+    D_anc_xy = D_anc_xy.union((x, y))
+
+    # second, construct the moralization of the subgraph
+    moral_G = nx.moralize_graph(G.subgraph(D_anc_xy))
+
+    # start BFS from X
+    marks = _bfs_with_marks(moral_G, x, z)
+
+    # if not all the Z is marked, then the set is not minimal
+    if any(node not in marks for node in z):
+        return False
+
+    # similarly, start BFS from Y and check the marks
+    marks = _bfs_with_marks(moral_G, y, z)
+    # if not all the Z is marked, then the set is not minimal
+    if any(node not in marks for node in z):
+        return False
+
+    return True
+
+
+def _bfs_with_marks(G, start_node, check_set):
```

Comment:
The `generic_bfs_edges` function does indeed do BFS. But there are so many possible nice additions to BFS that it can easily get bogged down if we try to put too much in.  Early on we had a beautiful version of BFS with hooks at every turn to allow the user to build just about any version they wanted. But all the hooks ended up making it slower and more importantly difficult to understand. The basic algorithm itself is not long or hard to read (as this code show). So, I would recommend leaving this private function here with the `marks` feature. I'm willing to bend on this if anyone would prefer moving it to the traversal subpackage.

Reply:
I moved `_bfs_with_marks` into `traversal` submodule for now. I think perhaps it could also be turned into a more `generic_bfs_edges_with_marks` like function if in the future, there is any algorithm internally that also needs this type of algorithm?